### ISSUE TITLE
Transactional endpoint not working behind proxy

### ISSUE THREAD

**Issue by albertoperdomo** [neo4j/neo4j#2985]
Hello, Alberto from GrapheneDB here.

One of our users is experiencing an issue where the URLs returned by the transaction endpoint are not honoring the SSL reverse proxy configuration, and reverting to `localhost:0` (local Neo4j config).

We have isolated this issue with 2.1.3 Community Edition. We haven't yet investigated if this issue is also present in older versions.

Example query to reproduce.

```
POST /db/data/transaction HTTP/1.1
Accept-Encoding: identity
Content-Length: 348
Host: appXXX.sb02.stations.graphenedb.com:24789
Content-Type: application/json
X-Stream: true;format=pretty
Authorization: Basic [[cut]]
User-Agent: py2neo/1.6.1 HTTPStream/1.1.0 Python/2.7.5-final-0 (linux2)
Connection: close.

{"statements":[{"statement":"start a=node:node_auto_index(name='6153') ,\n    b=node:node_auto_index(name='3597')\n    match 
p=a-[r*2..5]->b\n    with p, relationships(p) as rcoll\n    return p, REDUCE(totalTime=0, x in rcoll | totalTime + x.time) as 
totalTime\n    order by totalTime\n    LIMIT 5;","parameters":{},"resultDataContents":["REST"]}]}
```

The URLs in the response are generated with `localhost:0` instead of `appXXX.sb02.stations.graphenedb.com:24789`:

```
HTTP/1.1 201 Created
Location: http://appXXX.sb02.stations.graphenedb.com:24789/db/data/transaction/15
Content-Type: application/json
Access-Control-Allow-Origin: *
Connection: close
Server: Jetty(9.0.5.v20130815)

{"nodes":["http://localhost:0/db/data/node/390","http://localhost:0/db/data/node/425","http://localhost:0/db/data/node/487",
"http://localhost:0/db/data/node/481","http://localhost:0/db/data/node/472","http://localhost:0/db/data/node/391"],"length":5,
"relationships":["http://localhost:0/db/data/relationship/758","http://localhost:0/db/data/relationship/1121", [[cut]]
```

We guess that, when Py2Neo tries to load the data from the node, it tries to connect to `http://localhost:0`, which will always fail and yield connection refused error (which was the original issue submitted by the user).

The value `localhost:0` is used to configure Neo4j.

This problem seems to be related to the transaction API. For example, with a query like this works well:

```
GET /db/data/node/390 HTTP/1.1
Accept-Encoding: identity
Host: appXXX.sb02.stations.graphenedb.com:24789
X-Stream: true;format=pretty
Authorization: Basic [[cut]]
User-Agent: py2neo/1.6.1 HTTPStream/1.1.0 Python/2.7.5-final-0 (linux2)
Connection: close.
The response is generated with valid URLs:

HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
Access-Control-Allow-Origin: *
Content-Length: 1617
Server: Jetty(9.0.5.v20130815)

{
  "extensions" : {
  },
  "labels" : "http://appXXX.sb02.stations.graphenedb.com:24789/db/data/node/390/labels",
  "outgoing_relationships" : "http://appXXX.sb02.stations.graphenedb.com:24789/db/data/node/390/relationships/out",
  [[cut]]
```

**Comment by albertoperdomo** [comment]
@cleishm Seems this has not yet been backported to 2.1.x. Any plans on doing it?
Or will this be only addressed when 2.2.0 goes GA?

**Comment by jimwebber** [comment]
Hi @albertoperdomo,

With neo4j 2.1.7, I don't see this problem:

HTTP/1.1 201 Created
    Location: https://db-foo.graphenedb.com/db/data/transaction/8
    Content-Type: application/json
    Access-Control-Allow-Origin: *
    Content-Length: 1638
    Server: Jetty(9.0.5.v20130815)
    
    {"commit":"https://db-foo.graphenedb.com/db/data/transaction/8/commit","results":[{"columns":["n"],"data":[{"rest":[{"labels":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/labels","outgoing_relationships":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/relationships/out","traverse":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/traverse/{returnType}","all_typed_relationships":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/relationships/all/{-list|&|types}","self":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0","property":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/properties/{key}","properties":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/properties","outgoing_typed_relationships":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/relationships/out/{-list|&|types}","incoming_relationships":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/relationships/in","create_relationship":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/relationships","paged_traverse":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/paged/traverse/{returnType}{?pageSize,leaseTime}","all_relationships":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/relationships/all","incoming_typed_relationships":"https://db-foo.graphenedb.com/db/data/transactiondb/data/node/0/relationships/in/{-list|&|types}","metadata":{"id":0,"labels":["Person"]},"data":{"name":"Fred"}}]}]}],"transaction":{"expires":"Wed, 18 Mar 2015 11:51:38 +0000"},"errors":[]} 

I think that @jakewins PR was the final thing that fixed up base URIs (whereas mine did not). See: https://github.com/neo4j/neo4j/pull/3968

Can you verify that 2.1.7 does the right thing for you?

Jim

**Comment by albertoperdomo** [comment]
Sorry, I assumed that this had not been yet backported since the GH issue wasn't updated.
I just double-checked and confirm the fix is shipped with 2.1.7.

![Thanks](http://i.giphy.com/4lpctAAV9Azpm.gif)

Thanks @cleishm @jimwebber @jakewins ! I'm very happy to finally close this issue!!
